src/components/Shell/RoomShareDialog.tsx

```diff
@@ -0,0 +1,156 @@
+  const [isUnderstood, setIsUnderstood] = useState(false)
+  const handleClose = () => {
+    props.handleClose()
+    setPassword('')
+  }
+
+  useEffect(() => {
+    if (!isUnderstood) setPassword('')
+  }, [isUnderstood])
+
+  useEffect(() => {
+    if (!isAdvanced) setIsUnderstood(false)
+  }, [isAdvanced])
+
+  const url = window.location.href.split('#')[0]
+
+  const copyWithPass = async () => {
+    const secret = await encodePassword(props.roomId, password)
+    const params = new URLSearchParams()
+    params.set('secret', secret)
+    await props.copyToClipboard(
+      url + '#' + params.toString(),
+      'Private room URL with password copied to clipboard',
+      'warning'
+    )
+    handleClose()
+  }
+
+  const copyWithoutPass = async () => {
+    await props.copyToClipboard(
+      url,
+      'Private room URL without password copied to clipboard',
+      'success'
+    )
+    handleClose()
+  }
+
+  const showIf = (when: boolean) => {
+    return { display: when ? '' : 'none' }
+  }
+
+  return (
+    <Dialog
+      open={props.isOpen}
+      onClose={handleClose}
+      aria-labelledby="alert-dialog-title"
+      aria-describedby="alert-dialog-description"
+    >
+      <DialogTitle id="alert-dialog-title" sx={showIf(isAdvanced)}>
+        Copy URL with password
+        <Button onClick={() => setIsAdvanced(false)}>Simple</Button>
+      </DialogTitle>
+      <DialogContent sx={showIf(isAdvanced)}>
+        <DialogContentText sx={{ mb: 2 }}>
+          Copy URL to this private room containing an indecipherable hash of the
+          password. When using this URL, users will not need to enter the
+          password themselves.
+        </DialogContentText>
+        <Alert severity="error" sx={{ mb: 2 }}>
+          Be careful where and how this URL is shared. Anybody who obtains it
+          can enter the room. The sharing medium must be trusted, as well as all
+          potential recipients of the URL, just as if you were sharing the
+          password itself.
+        </Alert>
+        <Alert severity="warning">
+          By design, the password hash does not leave the web browser when this
+          URL is used to access the room. However, web browsers can still
+          independently record the full URL in the address history, and may even
+          store the history in the cloud if configured to do so.
+        </Alert>
+        <FormControlLabel
+          label="I understand the risks"
+          control={
+            <Checkbox
+              checked={isUnderstood}
+              onChange={e => setIsUnderstood(e.target.checked)}
+            />
+          }
+        />
+        <TextField
+          autoFocus
+          margin="none"
+          id="password"
+          label="Password"
+          type="password"
+          fullWidth
+          variant="standard"
+          value={password}
+          disabled={!isUnderstood}
+          onChange={e => setPassword(e.target.value)}
+        />
```

Comment:
Would it be possible to avoid prompting the intiating user to re-enter the password here? If they don't enter a matching password, URL recipients won't be able to join the room and the experience would be confusing for both parties.

Reply:
It is certainly possible, but there are advantages to this approach. 

Requiring reentry of the password makes it harder to join the room from another device if somebody who wasn't invited gains temporary access to a member's device. 

Requiring password reentry means that the intruder would only be able to talk in the room without suspicion for as long as they have the device in their possession without the owner knowing about it.

If the password reentry is removed, then only seconds are needed to fully compromise the room without any way to trace. 

Given that this is designated as an advanced feature not to be used willy nilly, perhaps this small barrier to its use is not such a bad thing.

I'm not so sure I would want to weaken the security of the room for a relatively niche feature like this. I mainly see it being valuable in allowing people to join easily, not necessarily to make it easier to share (beyond making it easier to get invited people to join).

Reply:
That makes sense. Thanks for sharing your thinking about this! I'm still concerned about scenarios where the sharer inputs the wrong password in this second prompt and then nobody can connect. While that's good for security, it would result in a confusing and frustrating UX. What do you think about treating a password mismatch as an error and prompting the user to correct their input before the link can be shared with others?

Reply:
That sounds like a good idea, I'll probably have time tomorrow to implement this and the other bits and pieces.